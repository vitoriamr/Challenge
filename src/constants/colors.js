export const LIGHT_GRAY = "#eaecf2";
export const LINE_GRAY = "#eeeeee";
