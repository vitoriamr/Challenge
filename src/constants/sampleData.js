export const DEFAULT_VALUE = "Select an Option...";

export const STATES = {
  NY: "New York",
  CA: "California",
  TX: "Texas"
};

export const CITIES = {
  NY_CITY: "New York City",
  BUFFALO: "Buffalo",
  LA: "Los Angeles",
  SF: "San Francisco",
  AUSTIN: "Austin",
  SA: "San Antonio"
};
