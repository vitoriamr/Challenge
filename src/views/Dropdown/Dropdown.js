import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import { LIGHT_GRAY, LINE_GRAY } from "../../constants/colors";
import dropdownArrowIcon from "../../icons/dropdownArrow.svg";

class Dropdown extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showDropdown: false
    };

    this.showDropdown = this.showDropdown.bind(this);
    this.closeDropdown = this.closeDropdown.bind(this);
  }

  showDropdown(event) {
    event.preventDefault();

    this.setState({ showDropdown: true }, () => {
      document.addEventListener("click", this.closeDropdown);
    });
  }

  closeDropdown(event) {
    if (!this.dropdownMenu.contains(event.target)) {
      this.setState({ showDropdown: false }, () => {
        document.removeEventListener("click", this.closeDropdown);
      });
    }
  }

  render() {
    const {
      available,
      currentValue,
      valueOptions,
      handleDropdown
    } = this.props;
    return (
      <div>
        <FirstButton
          available={available}
          onClick={available ? this.showDropdown : ""}
        >
          <FlexDiv>
            <DropdownText>{currentValue}</DropdownText>
            <Icon
              src={dropdownArrowIcon}
              facingUp={this.state.showDropdown}
              alt="dropdownArrowIcon"
            />
          </FlexDiv>
        </FirstButton>
        {this.state.showDropdown ? (
          <Options
            className="dropdown"
            ref={element => {
              this.dropdownMenu = element;
            }}
          >
            {valueOptions.map(option => (
              <div key={`value-${option}`}>
                <OptionsButton
                  key={`button-${option}`}
                  id={`button-${option}`}
                  onClick={event => {
                    handleDropdown(event, option);
                    this.setState({ showDropdown: false }, () => {
                      document.removeEventListener("click", this.closeDropdown);
                    });
                  }}
                >
                  <MenuText id={option}>{option}</MenuText>
                </OptionsButton>
              </div>
            ))}
          </Options>
        ) : null}
      </div>
    );
  }
}

Dropdown.propTypes = {
  available: PropTypes.bool.isRequired,
  currentValue: PropTypes.string.isRequired,
  valueOptions: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  handleDropdown: PropTypes.func.isRequired
};

const FirstButton = styled.button`
  padding: 0;
  border-radius: 4px;
  background: ${LIGHT_GRAY};
  margin-right: -6px;
  width: 155px;
  height: 30px;
  ${props => (props.available ? "" : "opacity: 0.3;")}
`;
FirstButton.displayName = "FirstButton";

const FlexDiv = styled.div`
  display: flex;
`;

const Icon = styled.img`
  margin-top: 1px;
  ${props => (props.facingUp ? "transform: rotateX(-180deg);" : "")}
  left: 86%;
  position: relative;
`;
Icon.displayName = "Icon";

const DropdownText = styled.div`
  font-size: 14px;
  line-height: 1.75;
  margin: -3px 0 0 9px;
  position: absolute;
`;
DropdownText.displayName = "DropdownText";

const Options = styled.div`
  width: 155px;
  height: auto;
  padding-top: 10px;
  text-align: center;
  border-radius: 3px;
  box-shadow: 0 10px 16px 0 rgba(0, 0, 0, 0.05);
`;
Options.displayName = "Options";

const OptionsButton = styled.button`
  padding: 0;
  border: none;
  background: transparent;
  width: 100%;
`;
OptionsButton.displayName = "OptionsButton";

const MenuText = styled.div`
  font-size: 14px;
  line-height: 1.75;

  &:hover {
    background-color: ${LINE_GRAY};
  }
`;
MenuText.displayName = "MenuText";

export default Dropdown;
