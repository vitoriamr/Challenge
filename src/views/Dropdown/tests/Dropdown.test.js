import React from "react";
import { shallow, mount } from "enzyme";

import { DEFAULT_VALUE, STATES, CITIES } from "../../../constants/sampleData";
import Dropdown from "../Dropdown";

describe("Dropdown", () => {
  const requiredProps = {
    available: true,
    currentValue: DEFAULT_VALUE,
    valueOptions: [STATES.CA, STATES.NY, STATES.TX],
    handleDropdown: () => {}
  };

  it("matches previous snapshot", () => {
    const wrapper = shallow(<Dropdown {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render default dropdown text", () => {
    const wrapper = shallow(<Dropdown {...requiredProps} />);
    expect(wrapper.find("DropdownText").text()).toBe(DEFAULT_VALUE);
  });

  it("opens dropdown options on click", () => {
    const wrapper = mount(<Dropdown {...requiredProps} />);
    wrapper.find("FirstButton").simulate("click");
    expect(wrapper.find("Options").exists()).toBe(true);
  });

  it("should call handleDropdown when selecting an option", () => {
    const props = {
      ...requiredProps,
      handleDropdown: jest.fn()
    };
    const wrapper = mount(<Dropdown {...props} />);
    jest.clearAllMocks();
    wrapper.find("FirstButton").simulate("click");
    wrapper.find(`OptionsButton[id="button-${STATES.NY}"]`).simulate("click");
    expect(props.handleDropdown).toHaveBeenCalledTimes(1);
  });

  it("should close dropdown when selecting an option", () => {
    const props = {
      ...requiredProps,
      handleDropdown: jest.fn()
    };
    const wrapper = mount(<Dropdown {...props} />);
    jest.clearAllMocks();
    wrapper.find("FirstButton").simulate("click");
    wrapper.find(`OptionsButton[id="button-${STATES.NY}"]`).simulate("click");
    expect(wrapper.find("Options").exists()).toBe(false);
  });

  describe("first dropdown", () => {
    it("should render New York label", () => {
      const wrapper = mount(<Dropdown {...requiredProps} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${STATES.NY}"]`).text()).toBe(
        STATES.NY
      );
    });

    it("should render California label", () => {
      const wrapper = mount(<Dropdown {...requiredProps} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${STATES.CA}"]`).text()).toBe(
        STATES.CA
      );
    });

    it("should render Texas label", () => {
      const wrapper = mount(<Dropdown {...requiredProps} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${STATES.TX}"]`).text()).toBe(
        STATES.TX
      );
    });
  });

  describe("second dropdown", () => {
    const props = {
      ...requiredProps,
      valueOptions: [
        CITIES.AUSTIN,
        CITIES.BUFFALO,
        CITIES.LA,
        CITIES.NY_CITY,
        CITIES.SA,
        CITIES.SF
      ]
    };
    it("should render Austin label", () => {
      const wrapper = mount(<Dropdown {...props} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${CITIES.AUSTIN}"]`).text()).toBe(
        CITIES.AUSTIN
      );
    });

    it("should render Buffalo label", () => {
      const wrapper = mount(<Dropdown {...props} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${CITIES.BUFFALO}"]`).text()).toBe(
        CITIES.BUFFALO
      );
    });

    it("should render Los Angeles label", () => {
      const wrapper = mount(<Dropdown {...props} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${CITIES.LA}"]`).text()).toBe(
        CITIES.LA
      );
    });

    it("should render New York City label", () => {
      const wrapper = mount(<Dropdown {...props} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${CITIES.NY_CITY}"]`).text()).toBe(
        CITIES.NY_CITY
      );
    });

    it("should render San Antonio label", () => {
      const wrapper = mount(<Dropdown {...props} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${CITIES.SA}"]`).text()).toBe(
        CITIES.SA
      );
    });

    it("should render San Francisco label", () => {
      const wrapper = mount(<Dropdown {...props} />);
      wrapper.find("FirstButton").simulate("click");
      expect(wrapper.find(`MenuText[id="${CITIES.SF}"]`).text()).toBe(
        CITIES.SF
      );
    });
  });
});
