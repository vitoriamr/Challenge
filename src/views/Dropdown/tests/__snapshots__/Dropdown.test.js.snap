// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches previous snapshot 1`] = `
<div>
  <FirstButton
    available={true}
    onClick={[Function]}
  >
    <styled.div>
      <DropdownText>
        Select an Option...
      </DropdownText>
      <Icon
        alt="dropdownArrowIcon"
        facingUp={false}
        src=""
      />
    </styled.div>
  </FirstButton>
</div>
`;
